src/cpp/_pythia8.cpp

```diff
@@ -17,9 +17,12 @@ using namespace pybind11::literals;
 float charge_from_pid(const ParticleData &pd, int pid)
 {
     auto pptr = pd.findParticle(pid);
-    assert(pptr); // never nullptr if charge_from_pid is used on particles produced by Pythia
-    // ParticleData returns partice even if anti-particle pid is used
-    return pid == pptr->id() ? pptr->charge() : -pptr->charge();
+    if (pptr) {
+        // ParticleData returns partice even if anti-particle pid is used
+        return pid == pptr->id() ? pptr->charge() : -pptr->charge();
+    } else {
+        return 0;
+    }    
 }
```

Comment:
Please revert. You cannot just return anything when the pid is not known to Pythia. This is hiding that the result is arbitrary. How do you know that charge = 0 is correct? If you don't like the assert, then throw an exception and then handle the exception in the client code which may have more information.

Reply:
Ok, I changed `charge_from_pid` so that it throws an exception.

The problem with exceptions is that the function charge stumbles upon unknown pdg id, while it is needed that it will return the values for known pdgs and inform which items are not known. I was trying to pass that info through exceptions, but pybind11 doesn't allow in a simple way to pass custom exception which contain additional information.

So I added new function `charge_nothrow` which returns 2 arrays: one for charge values and one boolean array whether that value is invalid. It is done for efficiency reasons, as one array with sentinel values (777) is enough to calculate boolean numpy array from it.

Reply:
Yep, `NaN` is more elegant. I changed the function to return NaN.